### ISSUE TITLE
[Bug][Numpy] Very basic functionality of Gluon is broken

### ISSUE THREAD

**Issue by sxjscience** [apache/incubator-mxnet#18193]
After https://github.com/apache/incubator-mxnet/pull/17283, the following very basic example of gluon is broken:

```python
import mxnet as mx
mx.npx.set_np()
from mxnet.gluon import nn
net = nn.Dense(16, in_units=16)
net.initialize()
out = net(mx.np.ones((32, 16)))
out.asnumpy()
```

```
~/mxnet/python/mxnet/numpy/random.py in uniform(low, high, size, dtype, ctx, out)
    135     inequality condition.
    136     """
--> 137     return _mx_nd_np.random.uniform(low, high, size=size, ctx=ctx, dtype=dtype, out=out)
    138 
    139 

~/mxnet/python/mxnet/ndarray/numpy/random.py in uniform(low, high, size, dtype, ctx, out)
    135     if size == ():
    136         size = None
--> 137     return _api_internal.uniform(low, high, size, ctx, dtype, out)
    138 
    139 

~/mxnet/python/mxnet/_ffi/_ctypes/function.py in __call__(self, *args)
    113                 self.handle, values, tcodes, ctypes.c_int(num_args),
    114                 ctypes.byref(ret_val), ctypes.byref(ret_tcode)) != 0:
--> 115             raise get_last_ffi_error()
    116         _ = temp_args
    117         _ = args

MXNetError: Traceback (most recent call last):
  File "../src/imperative/./imperative_utils.h", line 251
MXNetError: Check failed: outputs[i]->dtype() == out_types[i] (0 vs. 1) : 0-th output has invalid dtype. Expecting 1 got 0 in operator _npi_uniform
```

**Comment by sxjscience** [comment]
@yzhliu @JiangZhaoh @szha

**Comment by yzhliu** [comment]
Assignee: @JiangZhaoh

**Comment by JiangZhaoh** [comment]
Sorry for that. I'll fix it as soon as possible.

**Comment by szha** [comment]
which tests were supposed to catch this issue?

**Comment by sxjscience** [comment]
@szha For example, https://github.com/apache/incubator-mxnet/blob/73d1b055d04a0f0f511a9cc2dd46ae2eb03a8628/tests/python/unittest/test_numpy_gluon.py#L382-L411 I think there are tests in `test_numpy_gluon.py` that will capture this. Let me confirm.

**Comment by JiangZhaoh** [comment]
That's my issue. In order to not effect original tests, I didn't change the decorator "use_np", which still use `return use_np_shape(use_np_array(func))`. But in npx.set_np() I default flag `dtype` as `True` and then `set_np_default_dtype(dtype)`.
Details in /mxnet/python/mxnet/util.py

**Comment by sxjscience** [comment]
OK, so this is not related to CI.

**Comment by JiangZhaoh** [comment]
I have found the reason. It because `set_np()` only change the mxnet.numpy functions to default dtype as float64, but doesn't effect mxnet.ndarray functions which still default float32.
The issue mentioned above just like this code:
```
>>> import mxnet
>>> from mxnet import numpy as np
>>> from mxnet import ndarray as nd
>>> from mxnet import npx
>>> npx.set_np()
>>> a = nd.zeros(5)
>>> a.dtype
<class 'numpy.float32'>
>>> np.random.uniform(low = -1, high = 1, size = (5), out = a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/mxnet/python/mxnet/numpy/random.py", line 137, in uniform
    return _mx_nd_np.random.uniform(low, high, size=size, ctx=ctx, dtype=dtype, out=out)
  File "/home/ubuntu/mxnet/python/mxnet/ndarray/numpy/random.py", line 145, in uniform
    return _api_internal.uniform(low, high, size, ctx, dtype, out)
  File "mxnet/_ffi/_cython/./function.pxi", line 188, in mxnet._ffi._cy3.core.FunctionBase.__call__
  File "mxnet/_ffi/_cython/./function.pxi", line 133, in mxnet._ffi._cy3.core.FuncCall
  File "mxnet/_ffi/_cython/./base.pxi", line 90, in mxnet._ffi._cy3.core.CALL
mxnet.base.MXNetError: Traceback (most recent call last):
  File "../src/imperative/./imperative_utils.h", line 251
MXNetError: Check failed: outputs[i]->dtype() == out_types[i] (0 vs. 1) : 0-th output has invalid dtype. Expecting 1 got 0 in operator _npi_uniform
>>> npx.set_np(dtype = False)
>>> np.random.uniform(low = -1, high = 1, size = (5), out = a)
[0.09762704 0.18568921 0.43037868 0.6885315  0.20552671]
<NDArray 5 @cpu(0)>
```
So, I think I'd better not to set `dtype=True` flag in `npx.set_np()`, seperate this flag along in a single function `npx.set_numpy_default_dtype()` instead. Or if there any other better solutions, please tell me.

**Comment by sxjscience** [comment]
I think one of the root cause is that gluon optimizer and initializer are still using the legacy NDArray. We should fix that.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by yzhliu** [comment]
should be resolved by #18200
